### DISCUSSION TITLE
Failed to fetch Sample from MONAI Label Server

### DISCUSSION THREAD

**Discussion by dyollb** [Project-MONAI/MONAILabel#858] [Q&A] [answered]
I am trying MONAILabel for the first time.

## I put my images in following hierarchy (I have about 60 images, and 3 labeled images, single label):
```
- F:\Data\eyes_MONAILabel
  - labels
    - final
      - prefix_001.nii.gz
      - prefix_002.nii.gz
      - prefix_003.nii.gz
  - prefix_001.nii.gz
  - prefix_002.nii.gz
  - prefix_003.nii.gz
  - prefix_004.nii.gz
  - prefix_005.nii.gz
  - ...
```

## I installed
- slicer preview 5.1.0-2022-06-28, and MONAI plugin
- MONAILabel current main branch (`pip install monai-label` did not work, when I entered the URL in Slicer I got an error)
  - git clone https://github.com/Project-MONAI/MONAILabel
  - pip install torch torchvision --extra-index-url https://download.pytorch.org/whl/cu113
  - pip install -e MONAILabel

## Starting the server
- first downloed the app
- `monailabel apps --download --name radiology --output apps`
- the start the server
- `monailabel start_server --app apps/radiology --studies "F:\Data\eyes_MONAILabel" --conf models deepedit`

## The problem I am having
- I could start the plugin and enter the URL. 
- Then I ran `Train` until it reached 100%
- Then I pressed `Nex Sample`
![image](https://user-images.githubusercontent.com/12702862/176842142-a7b8f468-3302-4e81-b852-61bb2a80f8b7.png)

Since the error pop-up cannot be resized I scrolled down and show the rest of the callstack:
![image](https://user-images.githubusercontent.com/12702862/176843948-c6d9bfbe-e208-41f9-b243-36eea69d0712.png)

Is this due to incompatible Slicer / Slicer plugin / MONAILabel app / MONAILabel package versions?

**Comment by dyollb** [comment]
Does the info from http://127.0.0.1:8000/info/ make sense? I am trying to segment a single label in this example (eyes). If not, where do I tell deepedit to train for a single label?
```
{
    "name": "MONAILabel - Radiology",
    "description": "DeepLearning models for radiology",
    "version": "2.0",
    "labels": [
        "background",
        "spleen",
        "right kidney",
        "left kidney",
        "liver",
        "stomach",
        "aorta",
        "inferior vena cava"
    ],
    "models": {
        "deepedit": {
            "type": "deepedit",
            "labels": {
                "spleen": 1,
                "right kidney": 2,
                "left kidney": 3,
                "liver": 6,
                "stomach": 7,
                "aorta": 8,
                "inferior vena cava": 9,
                "background": 0
            },
            "dimension": 3,
            "description": "A DeepEdit model for volumetric (3D) segmentation over 3D Images",
            "config": {}
        },
        "deepedit_seg": {
            "type": "segmentation",
            "labels": {
                "spleen": 1,
                "right kidney": 2,
                "left kidney": 3,
                "liver": 6,
                "stomach": 7,
                "aorta": 8,
                "inferior vena cava": 9,
                "background": 0
            },
            "dimension": 3,
            "description": "A DeepEdit model for volumetric (3D) segmentation over 3D Images",
            "config": {}
        },
        "Histogram+GraphCut": {
            "type": "scribbles",
            "labels": {
                "spleen": 1,
                "right kidney": 2,
                "left kidney": 3,
                "liver": 6,
                "stomach": 7,
                "aorta": 8,
                "inferior vena cava": 9,
                "background": 0
            },
            "dimension": 3,
            "description": "A post processing step with histogram-based GraphCut for Generic segmentation",
            "config": {
                "num_bins": 64,
                "lamda": 1.0,
                "sigma": 0.1
            }
        },
        "GMM+GraphCut": {
            "type": "scribbles",
            "labels": {
                "spleen": 1,
                "right kidney": 2,
                "left kidney": 3,
                "liver": 6,
                "stomach": 7,
                "aorta": 8,
                "inferior vena cava": 9,
                "background": 0
            },
            "dimension": 3,
            "description": "A post processing step with GMM-based GraphCut for Generic segmentation",
            "config": {
                "num_mixtures": 20,
                "lamda": 5.0,
                "sigma": 0.5
            }
        }
    },
    "trainers": {
        "deepedit": {
            "description": "Train DeepEdit model for 3D Images",
            "config": {
                "name": "train_01",
                "pretrained": 1,
                "device": "cuda",
                "max_epochs": 50,
                "early_stop_patience": -1,
                "val_split": 0.2,
                "train_batch_size": 1,
                "val_batch_size": 1,
                "multi_gpu": true,
                "gpus": "all",
                "dataset": [
                    "CacheDataset",
                    "PersistentDataset",
                    "SmartCacheDataset",
                    "Dataset"
                ],
                "dataloader": [
                    "ThreadDataLoader",
                    "DataLoader"
                ]
            }
        }
    },
    "strategies": {
        "random": {
            "description": "Random Strategy"
        },
        "first": {
            "description": "Get First Sample"
        }
    },
    "scoring": {},
    "train_stats": {
        "deepedit": {
            "total_epochs": 50,
            "total_iterations": 2,
            "epoch": 50,
            "start_ts": 1656658276,
            "total_time": "0:07:34",
            "best_metric": 0.8084278702735901,
            "train": {
                "metrics": {
                    "train_dice": 0.9283537268638611,
                    "spleen_dice": 0.9283537268638611,
                    "right kidney_dice": 0.0,
                    "left kidney_dice": 0.0,
                    "liver_dice": 0.0,
                    "stomach_dice": 0.0,
                    "aorta_dice": 0.0,
                    "inferior vena cava_dice": 0.0
                },
                "key_metric_name": "train_dice",
                "best_metric": 0.9344421625137329,
                "best_metric_epoch": 47
            },
            "eval": {
                "metrics": {
                    "val_mean_dice": 0.6395465135574341,
                    "spleen_dice": 0.6395465135574341,
                    "right kidney_dice": 0.0,
                    "left kidney_dice": 0.0,
                    "liver_dice": 0.0,
                    "stomach_dice": 0.0,
                    "aorta_dice": 0.0,
                    "inferior vena cava_dice": 0.0
                },
                "key_metric_name": "val_mean_dice",
                "best_metric": 0.8084278702735901,
                "best_metric_epoch": 28
            }
        }
    },
    "datastore": {
        "total": 64,
        "completed": 3,
        "label_tags": {
            "final": 3
        }
    }
}
```

**Comment by diazandr3s** [comment, marked as answer]
This issue is being addressed here https://github.com/Project-MONAI/MONAILabel/issues/859 and here https://github.com/Project-MONAI/MONAILabel/issues/852#issuecomment-1171105619

**Comment by diassio** [comment]
Hello! I also have an issue as Failed to fetch Sample from MONAI Label Server, however a different origin. 
<img width="415" alt="Screenshot 2023-01-12 at 18 25 31" src="https://user-images.githubusercontent.com/96193983/212066012-e4834bec-57e4-4166-83bf-7aa452ad0d43.png">

**Reply by diazandr3s** [reply]
Hi @diassio,

Thanks for sharing your experience.
Which dataset are you using? Can you provide more details about this? Can you share the logs you're getting?

**Reply by diassio** [reply]
Hi!
I'm using custom/private brain CT dataset for ich segmentation. I've found, that some cases form my dataset could not be loaded to the 3d slicer causing "Failed to fetch Sample from MONAI Label Server" error. If I try to load next sample, it may load perfectly normal and ready for segmentation, but "Run" button becomes unclickable so I have to reload 3d Slicer. 
Logs are unremarkable. 
Also, I've tried to load those "broken" cases manually to 3D Slicer and found that they cause this error:
<img width="421" alt="Screenshot 2023-01-13 at 11 12 32" src="https://user-images.githubusercontent.com/96193983/212242497-e422a41f-f4a2-4de7-b6ac-d6bb24c5891c.png">

**Reply by diazandr3s** [reply]
Hi @diassio,

Thanks for the update.
It seems those files you can't load in Slicer are corrupted or have missing/wrong NIfTI headers. I don't think this is a MONAI Label issue. Can you check that?
The reason I suggest you share the MONAI Label logs is that the print screen you've shared says very little information about the issue.

**Reply by diassio** [reply]
Yes, I guess it is not from the MONAI Label side. It's just 3D slicer can't open some files. It is kinda strange, because those files could be opened with `nibabel.load()` perfectly fine

**Reply by diazandr3s** [reply]
I think Slicer uses itk library to manage the images.

**Reply by lassoan** [reply]
NIFTI is a very poorly defined file format, with lots of room for ambiguities and inconsistencies. If you create a NIFTI file using one tool and you read it with another then it may fail most often due to numerical inaccuracies and inconsistencies in image orientation stored in SFORM and QFORM. nibabel does not need to reconstruct a 3D volume from a NIFTI file, so it is usually less sensitive to image geometry issues compared to ITK. I would recommend to post an example image to the [ITK forum](https://discourse.itk.org/) so that they can investigate the issue and/or you can fix up your image headers to remove inaccuracies/inconsistencies (make sure that only SFORM or QFORM is used - not both; and if you use SFORM then have exactly orthogonal image axes).

**Reply by diassio** [reply]
okay, thank you for reply!
